_posts/2016-06-30-state-in-scalable-architectures.md

```diff
@@ -0,0 +1,173 @@
+
+### Object Oriented Programming State
+
+>> In most forms of object-oriented programming (OOP) an object is seen as consisting of some state together with a set of procedures for accessing and manipulating that state.
+
+As [Uncle Bob](https://www.wikiwand.com/en/Robert_Cecil_Martin) said once, programming paradigms succeeds, paradoxically, by removing power. Java succeed because it removed from developers the power to manage memory. OOP strongly suggests that code that mutates state should be colocated with that state. That guarantee, along preconditions and postconditions, makes code much easier to reason about and test.
+
+Using inheritance to share code is a bad idea based on those premises. State mutators will be located in different files, even if they finally constitute a single object on runtime, obscuring the readability of the code.
+
+### Functional Programming State
+
+FP tries to solve the complexity problem avoiding mutable state. Thanks to that we can have functions whose outputs depend completely on their inputs. Avoiding is not as simple as closing your eyes, though. There are some scenarios where we need to handle state. Let's use an [example](https://github.com/fpinscala/fpinscala/blob/master/answers/src/main/scala/fpinscala/state/State.scala) from [Functional programming in Scala book](https://www.manning.com/books/functional-programming-in-scala):
+
+```scala
+  case class Simple(seed: Long) extends RNG {
+    def nextInt: (Int, RNG) = {
+      val newSeed = (seed * 0x5DEECE66DL + 0xBL) & 0xFFFFFFFFFFFFL
+      val nextRNG = Simple(newSeed)
+      val n = (newSeed >>> 16).toInt
+      (n, nextRNG)
+    }
+  }
+```
+
+This random numbers generator class depends on the previous seed to generate new random numbers. The OOP approach would be keeping the old seed as state of an object and mutating it every time that `nextInt` is called. FP approach encapsulates that seed in a tuple along the result. There is not position in memory that has been updated. Everything is immutable and the result of a function is deterministic depending on the input. That property is called [Referential Transparency](https://www.wikiwand.com/en/Referential_transparency) and we can see its power if we think in terms of composability and modularity.
+
+### Actor State
+
+[Akka](http://akka.io/) is a toolkit that implements the [Actor Model](https://www.wikiwand.com/en/Actor_model). This model handles state with a mix of OOP and FP. These are the three basic rules for Akka concurrency contained in [this post](http://letitcrash.com/post/28901663062/throttling-messages-in-akka-2):
+
+>> Rule 1: Each actor processes the messages from its inbox sequentially (not concurrently). So a clerk never picks two or more messages and works them side-by-side.
```

Comment:
It's a quote from the post that I'm linking previously, if the reader wants or needs more context they can go to that post.